core/src/main/scala/com/evolutiongaming/kafka/flow/PartitionFlow.scala

```diff
@@ -191,20 +191,21 @@ object PartitionFlow {
       _ <- log.debug(s"finished processing records")
     } yield ()
 
+    def triggerTimersForKey(key: String, timestamp: Timestamp): F[Unit] =
+      cache.get(key).flatMap(_.traverse_(state => state.timers.set(timestamp) >> state.timers.trigger(state.flow)))
+
     def triggerTimers = for {
       _ <- log.debug("triggering timers")
 
       clock     <- Clock[F].instant
       timestamp <- timestamp updateAndGet (_.copy(clock = clock))
 
       keys <- cache.keys
-      _ <- keys.toVector.parTraverse_ { key =>
```

Comment:
what's expected number of `keys` here?

Reply:
There's no expected number, it all depends on a specific application. Some might have hundreds, some might have hundreds of thousands

Reply:
JFYI: we were fixing problem of calling `parTraverse` / `parSequence` against huge collections and this is what helped:

```scala
    def parFoldMap2[F[_]: Monad: Parallel: Runtime, B: Monoid](
      f: A => F[B]
    ): F[B] = {
      if (self.isEmpty) {
        Monoid[B].empty.pure[F]
      } else {

        def fold(self: Iterable[A]) = {
          self.foldLeft(Monoid[B].empty.pure[F]) {
            case (b, a) =>
              for {
                b <- b
                a <- f(a)
              } yield {
                b.combine(a)
              }
          }
        }

        Runtime[F]
          .availableCores
          .flatMap { cpus =>
            val size = self.size / cpus.max(1)
            if (size > 0) {
              self
                .grouped(size)
                .parFoldMap1(fold)
            } else {
              fold(self)
            }
          }
      }
    }

```

Reply:
this is for the case when the work we parallelise is mostly CPU bound

Reply:
btw, not sure whether you checked, but `parTraverseN` is implemented via `parTraverse` + `semaphore`, hence it still produces huge number of fibers.

Reply:
in case of non cpu bound work - I'd suggest using `cpu * multiplier`, where you configure multiplier based on the workload you have…

Reply:
> parTraverseN is implemented via parTraverse + semaphore

Yes, I know about it. The number of fibers itself isn't an issue while they're inactive; it's the number of concurrently active fibers that try to execute some code; in timers' case it's user-defined `Tick` code (which we know nothing about, could be either CPU- or IO-intensive) plus `TimerFlow` code which is usually both CPU-intensive (encode snapshot as a JSON) and IO-intensive (write the result to the storage)

Reply:
While this would be convenient to configure the limit based on the number of CPUs for CPU-intensive part of that (JSON encoding, for example), this may be not as good for IO-intensive part of work, namely Cassandra interactions, where I'd prefer to have a steady and a predictable load profile that doesn't depend on a number of CPUs

Reply:
sure, you are the maintainer :)